packages/core/contracts/oracle/implementation/VotingV2.sol

```diff
@@ -920,7 +921,23 @@ contract VotingV2 is
             voterStake.lastRequestIndexConsidered = requestIndex + 1;
         }
 
-        if (slash != 0) applySlashToVoter(slash, voterStake, voterAddress);
+        // If there is any remaining slashing then apply it. This occurs when there is unapplied slashing in the loop
+        // due to the last unlashed elements all being all from the same round. i.e we only slash within the loop when
+        // transitioning between rounds and the last round is slashed here. Note that there is a special case that needs
+        // to be considered separately: if the indexTo is less than the priceRequestIndex.length then we know we
+        // are dealing with a partial index slashing via a call to updateTrackersRange. If this is the case, we need to
+        // be careful for ranges that bisect a voting round as all slashing must be applied linearly and independently
+        // within each round. If we are in this case and the element before the indexTo (last element executed in the
+        // loop) has the same voting round as the indexTo (element after the last element in the loop) then we know that
+        // we've bisected a round and should store the unapplied slashing which will seed this method on the next entry
+        // such that the slashing will be applied linearly, not compounding with other slashing within the same round.
+        if (slash != 0)
+            if (
+                indexTo < priceRequestIds.length &&
+                priceRequests[priceRequestIds[indexTo - 1]].lastVotingRound ==
```

Comment:
Will rolled votes cause us to apply _too often_? For instance, if `i` and `i + 1` are both in the same round, `i` is rolled, and indexTo is `i + 1`, will we end up applying the slash despite there not being a change in round? This might be related to the indexTo++ line above, so it would be great to understand that a little better and how it solves for cases like two or three rolled votes in a row, etc.

Reply:
this was a great point! I've now addressed this due to adding in some changes that @md0x did in [this](https://github.com/UMAprotocol/protocol/pull/4064) pr. The key logic that avoids this now is this:
```
if (deletedRequests[requestIndex] != 0) {
                requestIndex = deletedRequests[requestIndex];
                continue;
            }
```

Such that if there are sequentially rolled requests we re-enter and increment to not double apply slashing. I've added a unit test to this effect.

note the `indexTo++` logic was also refined, please see comment to this effect above.